src/sinks/util/sink.rs

```diff
@@ -364,6 +386,26 @@ where
                 continue;
             }
 
+            // Cleanup of in flight futures
+            if let Some(in_flight) = this.in_flight.as_mut() {
+                if in_flight.len() > this.partitions.len() {
+                    // There is at least one in flight future without a partition to check it
+                    // so we will do it here.
+                    let mut remove = Vec::new();
+                    for (partition, req) in in_flight.iter_mut() {
+                        if !this.partitions.contains_key(partition)
+                            && matches!(req.poll_unpin(cx), Poll::Ready(()))
+                        {
+                            remove.push(partition.clone());
+                        }
+                    }
+
+                    for partition in remove {
+                        in_flight.remove(&partition);
+                    }
```

Comment:
Could this be accomplished with `retain` to avoid the clones, something like:
```suggestion
                    in_flight.retain(|partition, req| {
                        this.partitions.contains_key(partition)
                            || !matches!(req.poll_unpin(cx), Poll::Ready(()))
                    });
```